File: libs/sdk-python/src/daytona/internal/event_subscriber.py
# Copyright 2025 Daytona Platforms Inc.
# SPDX-License-Identifier: Apache-2.0

"""WebSocket event subscriber for real-time events via Socket.IO."""

from __future__ import annotations

import asyncio
import logging
import threading
from collections.abc import Awaitable
from typing import Any, Callable, Union

import socketio

logger = logging.getLogger(__name__)

# Handler receives (event_name, raw_data).
EventHandler = Callable[[str, Any], None]
AsyncEventHandler = Callable[[str, Any], Union[Awaitable[None], None]]


class AsyncEventSubscriber:
    """Async event subscriber that connects to the Socket.IO notification gateway."""

    _api_url: str
    _token: str
    _organization_id: str | None
    _sio: socketio.AsyncClient | None
    _connected: bool
    _failed: bool
    _fail_error: str | None
    _listeners: dict[str, list[AsyncEventHandler]]
    _registered_events: set[str]
    _disconnect_task: asyncio.Task[None] | None
    _subscription_timers: dict[str, threading.Timer]
    _subscription_ttls: dict[str, float]
    _disconnect_generation: int
    _lock: asyncio.Lock

    _DISCONNECT_DELAY: float = 30.0

    def __init__(self, api_url: str, token: str, organization_id: str | None = None):
        self._api_url = api_url
        self._token = token
        self._organization_id = organization_id
        self._sio = None
        self._connected = False
        self._failed = False
        self._fail_error = None
        self._listeners = {}
        self._registered_events = set()
        self._disconnect_task = None
        self._subscription_timers = {}
        self._subscription_ttls = {}
        self._connect_task: asyncio.Task[None] | None = None
        self._disconnect_generation = 0
        self._lock = asyncio.Lock()

    def ensure_connected(self) -> None:
        """Idempotent: ensure a connection attempt is in progress or already established.

        Non-blocking. Creates a background task if not already connected and no
        attempt is currently running.
        """
        if self._connected:
            return
        if self._connect_task is not None and not self._connect_task.done():
            return

        async def _connect() -> None:
            try:
                await self.connect()
            except Exception:
                pass  # Callers check is_connected when they need it

        try:
            loop = asyncio.get_running_loop()
        except RuntimeError:
            pass  # No event loop — will connect on first await
        else:
            self._connect_task = loop.create_task(_connect())

    async def connect(self, timeout: float = 5.0) -> None:
        """Establish the Socket.IO connection. Raises on failure."""
        async with self._lock:
            if self._connected:
                return
            old_sio = self._sio
            self._sio = None

        if old_sio:
            await old_sio.disconnect()

        origin = self._api_url.rstrip("/")
        if origin.endswith("/api"):
            origin = origin[:-4]

        sio = socketio.AsyncClient(
            reconnection=True,
            reconnection_attempts=0,
            reconnection_delay=1,
            reconnection_delay_max=30,
            logger=False,
            engineio_logger=False,
        )

        async with self._lock:
            self._sio = sio

        subscriber = self

        @sio.event  # type: ignore[misc]
        async def connect() -> None:  # pyright: ignore[reportUnusedFunction]
            subscriber._connected = True
            subscriber._failed = False
            subscriber._fail_error = None

        @sio.event  # type: ignore[misc]
        async def disconnect() -> None:  # pyright: ignore[reportUnusedFunction]
            subscriber._connected = False

        @sio.event  # type: ignore[misc]
        async def connect_error(data: Any) -> None:  # pyright: ignore[reportUnusedFunction]
            subscriber._connected = False
            subscriber._failed = True
            subscriber._fail_error = f"WebSocket connection failed: {data}"

        # Re-register any events that were added before the socket was created
        async with self._lock:
            pending_events = list(self._registered_events)
            self._registered_events.clear()
            self._register_events(pending_events)

        connect_url = origin
        if self._organization_id:
            connect_url = f"{origin}?organizationId={self._organization_id}"

        try:
            await asyncio.wait_for(
                sio.connect(
                    connect_url,
                    socketio_path="/api/socket.io/",
                    auth={"token": self._token},
                    transports=["websocket"],
                    headers={},
                    wait=True,
                ),
                timeout=timeout,
            )
        except asyncio.TimeoutError:
            self._failed = True
            self._fail_error = "WebSocket connection timed out"
            raise ConnectionError(self._fail_error) from None
        except Exception as e:
            self._failed = True
            self._fail_error = f"WebSocket connection failed: {e}"
            raise ConnectionError(self._fail_error) from e

    def subscribe(
        self,
        resource_id: str,
        handler: AsyncEventHandler,
        events: list[str],
        ttl_seconds: float = 0,
    ) -> Callable[[], None]:
        """Subscribe to specific events for a resource.

        Args:
            resource_id: The ID of the resource (e.g. sandbox ID).
            handler: Callback receiving (event_name, raw_data).
            events: List of Socket.IO event names to listen for (e.g. ["sandbox.state.updated", "sandbox.created"]).

        Returns:
            Unsubscribe function.
        """
        self.ensure_connected()

        self._disconnect_generation += 1

        if self._disconnect_task and not self._disconnect_task.done():
            _ = self._disconnect_task.cancel()
            self._disconnect_task = None

        # Register any new events with the Socket.IO client
        self._register_events(events)

        if resource_id not in self._listeners:
            self._listeners[resource_id] = []
        self._listeners[resource_id].append(handler)

        if ttl_seconds > 0:
            self._subscription_ttls[resource_id] = ttl_seconds
            self._start_subscription_timer(resource_id)
        else:
            self._clear_subscription_timer(resource_id)
            self._subscription_ttls.pop(resource_id, None)

        def unsubscribe() -> None:
            handlers = self._listeners.get(resource_id)
            if handlers and handler in handlers:
                handlers.remove(handler)
                if not handlers:
                    self._unsubscribe_resource(resource_id)

        return unsubscribe

    def refresh_subscription(self, resource_id: str) -> bool:
        if resource_id not in self._subscription_ttls:
            return False

        self._start_subscription_timer(resource_id)
        return True

    def _register_events(self, events: list[str]) -> None:
        """Register Socket.IO event handlers (idempotent — each event is registered once)."""
        subscriber = self

        for event_name in events:
            if event_name in self._registered_events:
                continue
            self._registered_events.add(event_name)

            # If socket isn't created yet, the event will be registered when connect() runs
            if not self._sio:
                continue

            def _make_handler(evt: str) -> Callable[..., Any]:
                async def _handler(data: Any) -> None:
                    resource_id = _extract_id_from_event(data)
                    if resource_id:
                        await subscriber._dispatch(resource_id, evt, data)

                return _handler

            self._sio.on(event_name, _make_handler(event_name))  # pyright: ignore[reportUnusedCallResult]

    def _schedule_delayed_disconnect(self) -> None:
        generation = self._disconnect_generation

        async def _delayed() -> None:
            await asyncio.sleep(self._DISCONNECT_DELAY)
            if generation != self._disconnect_generation or self._listeners:
                return
            await self._disconnect(expected_generation=generation)

        try:
            self._disconnect_task = asyncio.get_event_loop().create_task(_delayed())
        except RuntimeError:
            pass

    @property
    def is_connected(self) -> bool:
        return self._connected

    @property
    def is_failed(self) -> bool:
        return self._failed

    @property
    def fail_error(self) -> str | None:
        return self._fail_error

    async def disconnect(self) -> None:
        """Disconnect and clean up resources."""
        await self._disconnect()

    async def _disconnect(self, expected_generation: int | None = None) -> None:
        if self._connect_task and not self._connect_task.done():
            self._connect_task.cancel()

        async with self._lock:
            if expected_generation is not None:
                if expected_generation != self._disconnect_generation or self._listeners:
                    return

            sio = self._sio
            self._sio = None
            self._clear_subscription_timers()
            self._listeners.clear()
            self._registered_events.clear()
            self._connected = False

        if sio:
            await sio.disconnect()

    async def _dispatch(self, resource_id: str, event_name: str, data: Any) -> None:
        for handler in list(self._listeners.get(resource_id, [])):
            try:
                result = handler(event_name, data)
                if result is not None and asyncio.iscoroutine(result):
                    await result
            except Exception:
                pass

    def _start_subscription_timer(self, resource_id: str) -> None:
        ttl_seconds = self._subscription_ttls.get(resource_id)
        if ttl_seconds is None or ttl_seconds <= 0:
            return

        self._clear_subscription_timer(resource_id)

        try:
            loop = asyncio.get_running_loop()
        except RuntimeError:
            loop = None

        current_timer: threading.Timer | None = None

        def _expire() -> None:
            if self._subscription_timers.get(resource_id) is not current_timer:
                return

            self._clear_subscription_timer(resource_id)
            self._subscription_ttls.pop(resource_id, None)
            if loop is not None and loop.is_running():
                loop.call_soon_threadsafe(self._unsubscribe_resource, resource_id)
            else:
                self._unsubscribe_resource(resource_id)

        current_timer = threading.Timer(ttl_seconds, _expire)
        current_timer.daemon = True
        self._subscription_timers[resource_id] = current_timer
        current_timer.start()

    def _clear_subscription_timer(self, resource_id: str) -> None:
        timer = self._subscription_timers.pop(resource_id, None)
        if timer is not None:
            timer.cancel()

    def _clear_subscription_timers(self) -> None:
        for timer in self._subscription_timers.values():
            timer.cancel()
        self._subscription_timers.clear()
        self._subscription_ttls.clear()

    def _unsubscribe_resource(self, resource_id: str) -> None:
        if resource_id in self._listeners:
            del self._listeners[resource_id]

        self._clear_subscription_timer(resource_id)
        self._subscription_ttls.pop(resource_id, None)

        if not self._listeners:
            self._schedule_delayed_disconnect()


class SyncEventSubscriber:
    """Sync event subscriber using socketio.Client on a background thread."""

    _api_url: str
    _token: str
    _organization_id: str | None
    _sio: socketio.Client | None
    _connected: bool
    _failed: bool
    _fail_error: str | None
    _listeners: dict[str, list[EventHandler]]
    _registered_events: set[str]
    _lock: threading.Lock
    _bg_thread: threading.Thread | None
    _disconnect_timer: threading.Timer | None
    _subscription_timers: dict[str, threading.Timer]
    _subscription_ttls: dict[str, float]
    _disconnect_generation: int
    _closed: bool

    _DISCONNECT_DELAY: float = 30.0

    def __init__(self, api_url: str, token: str, organization_id: str | None = None):
        self._api_url = api_url
        self._token = token
        self._organization_id = organization_id
        self._sio = None
        self._connected = False
        self._failed = False
        self._fail_error = None
        self._listeners = {}
        self._registered_events = set()
        self._lock = threading.Lock()
        self._bg_thread = None
        self._disconnect_timer = None
        self._subscription_timers = {}
        self._subscription_ttls = {}
        self._disconnect_generation = 0
        self._closed = False

    def ensure_connected(self) -> None:
        """Idempotent: ensure a connection attempt is in progress or already established.

        Non-blocking. Starts a background thread if not already connected and no
        attempt is currently running.
        """
        if self._closed or self._connected:
            return
        if self._bg_thread is not None and self._bg_thread.is_alive():
            return

        def _connect() -> None:
            try:
                self.connect()
            except Exception:
                pass  # Callers check is_connected when they need it

        self._bg_thread = threading.Thread(target=_connect, daemon=True)
        self._bg_thread.start()

    def connect(self, timeout: float = 5.0) -> None:
        """Establish the Socket.IO connection. Raises on failure."""
        with self._lock:
            if self._closed or self._connected:
                return
            old_sio = self._sio
            self._sio = None

        if old_sio:
            old_sio.disconnect()

        origin = self._api_url.rstrip("/")
        if origin.endswith("/api"):
            origin = origin[:-4]

        sio = socketio.Client(
            reconnection=True,
            reconnection_attempts=0,
            reconnection_delay=1,
            reconnection_delay_max=30,
            logger=False,
            engineio_logger=False,
        )
        with self._lock:
            self._sio = sio

        connected_event = threading.Event()
        error_holder: list[str] = []
        subscriber = self

        @sio.event  # type: ignore[misc]
        def connect() -> None:  # pyright: ignore[reportUnusedFunction]
            subscriber._connected = True
            subscriber._failed = False
            subscriber._fail_error = None
            connected_event.set()

        @sio.event  # type: ignore[misc]
        def disconnect() -> None:  # pyright: ignore[reportUnusedFunction]
            subscriber._connected = False

        @sio.event  # type: ignore[misc]
        def connect_error(data: Any) -> None:  # pyright: ignore[reportUnusedFunction]
            subscriber._connected = False
            subscriber._failed = True
            subscriber._fail_error = f"WebSocket connection failed: {data}"
            error_holder.append(str(data))
            connected_event.set()

        try:
            # Re-register any events that were added before the socket was created
            with self._lock:
                pending_events = list(self._registered_events)
                self._registered_events.clear()
            self._register_events(pending_events)

            connect_url = origin
            if self._organization_id:
                connect_url = f"{origin}?organizationId={self._organization_id}"

            sio.connect(
                connect_url,
                socketio_path="/api/socket.io/",
                auth={"token": self._token},
                transports=["websocket"],
                headers={},
                wait=True,
                wait_timeout=int(timeout),
            )

            if self._closed:
                self._sio = None
                self._connected = False
                sio.disconnect()
                return
        except Exception as e:
            self._failed = True
            self._fail_error = f"WebSocket connection failed: {e}"
            raise ConnectionError(self._fail_error) from e

        if not self._connected:
            self._failed = True
            err = error_holder[0] if error_holder else "unknown error"
            self._fail_error = f"WebSocket connection failed: {err}"
            raise ConnectionError(self._fail_error)

    def subscribe(
        self,
        resource_id: str,
        handler: EventHandler,
        events: list[str],
        ttl_seconds: float = 0,
    ) -> Callable[[], None]:
        """Subscribe to specific events for a resource.

        Args:
            resource_id: The ID of the resource (e.g. sandbox ID).
            handler: Callback receiving (event_name, raw_data).
            events: List of Socket.IO event names to listen for.

        Returns:
            Unsubscribe function.
        """
        self.ensure_connected()

        if self._disconnect_timer:
            self._disconnect_timer.cancel()
            self._disconnect_timer = None

        # Register any new events with the Socket.IO client
        self._register_events(events)

        with self._lock:
            self._disconnect_generation += 1
            if resource_id not in self._listeners:
                self._listeners[resource_id] = []
            self._listeners[resource_id].append(handler)

            if ttl_seconds > 0:
                self._subscription_ttls[resource_id] = ttl_seconds
                self._start_subscription_timer_locked(resource_id)
            else:
                self._clear_subscription_timer_locked(resource_id)
                self._subscription_ttls.pop(resource_id, None)

        def unsubscribe() -> None:
            with self._lock:
                handlers = self._listeners.get(resource_id)
                if handlers and handler in handlers:
                    handlers.remove(handler)
                    if not handlers:
                        self._unsubscribe_resource_locked(resource_id)
                should_cleanup = not self._listeners

            if should_cleanup:
                generation = self._disconnect_generation
                self._disconnect_timer = threading.Timer(
                    self._DISCONNECT_DELAY,
                    self._delayed_disconnect,
                    args=(generation,),
                )
                self._disconnect_timer.daemon = True
                self._disconnect_timer.start()

        return unsubscribe

    def refresh_subscription(self, resource_id: str) -> bool:
        with self._lock:
            if resource_id not in self._subscription_ttls:
                return False

            self._start_subscription_timer_locked(resource_id)
            return True

    def _register_events(self, events: list[str]) -> None:
        """Register Socket.IO event handlers (idempotent — each event is registered once)."""
        subscriber = self

        with self._lock:
            for event_name in events:
                if event_name in self._registered_events:
                    continue
                self._registered_events.add(event_name)

                # If socket isn't created yet, the event will be registered when connect() runs
                if not self._sio:
                    continue

                def _make_handler(evt: str) -> Callable[..., Any]:
                    def _handler(data: Any) -> None:
                        resource_id = _extract_id_from_event(data)
                        if resource_id:
                            subscriber._dispatch(resource_id, evt, data)

                    return _handler

                self._sio.on(event_name, _make_handler(event_name))  # pyright: ignore[reportUnusedCallResult]

    def _delayed_disconnect(self, generation: int) -> None:
        self._disconnect(permanent=False, expected_generation=generation)

    @property
    def is_connected(self) -> bool:
        return self._connected

    @property
    def is_failed(self) -> bool:
        return self._failed

    @property
    def fail_error(self) -> str | None:
        return self._fail_error

    def disconnect(self) -> None:
        """Disconnect and clean up resources."""
        self._disconnect(permanent=True)

    def _disconnect(self, permanent: bool, expected_generation: int | None = None) -> None:
        sio = None

        with self._lock:
            if expected_generation is not None:
                if expected_generation != self._disconnect_generation or self._listeners:
                    return
            if permanent:
                self._closed = True
            if self._disconnect_timer is not None:
                self._disconnect_timer.cancel()
                self._disconnect_timer = None
            sio = self._sio
            self._sio = None
            self._connected = False
            self._clear_subscription_timers_locked()
            self._listeners.clear()
            self._registered_events.clear()

        if sio:
            sio.disconnect()

    def _dispatch(self, resource_id: str, event_name: str, data: Any) -> None:
        with self._lock:
            handlers = list(self._listeners.get(resource_id, []))
        for handler in handlers:
            try:
                handler(event_name, data)
            except Exception:
                pass

    def _start_subscription_timer_locked(self, resource_id: str) -> None:
        ttl_seconds = self._subscription_ttls.get(resource_id)
        if ttl_seconds is None or ttl_seconds <= 0:
            return

        self._clear_subscription_timer_locked(resource_id)

        current_timer: threading.Timer | None = None

        def _expire() -> None:
            self._expire_subscription(resource_id, current_timer)

        current_timer = threading.Timer(ttl_seconds, _expire)
        current_timer.daemon = True
        self._subscription_timers[resource_id] = current_timer
        current_timer.start()

    def _clear_subscription_timer_locked(self, resource_id: str) -> None:
        timer = self._subscription_timers.pop(resource_id, None)
        if timer is not None:
            timer.cancel()

    def _clear_subscription_timers_locked(self) -> None:
        for timer in self._subscription_timers.values():
            timer.cancel()
        self._subscription_timers.clear()
        self._subscription_ttls.clear()

    def _expire_subscription(self, resource_id: str, current_timer: threading.Timer | None) -> None:
        with self._lock:
            if self._subscription_timers.get(resource_id) is not current_timer:
                return
            self._unsubscribe_resource_locked(resource_id)
            should_cleanup = not self._listeners
            generation = self._disconnect_generation

        if should_cleanup:
            self._disconnect_timer = threading.Timer(self._DISCONNECT_DELAY, self._delayed_disconnect, args=(generation,))
            self._disconnect_timer.daemon = True
            self._disconnect_timer.start()

    def _unsubscribe_resource_locked(self, resource_id: str) -> None:
        self._listeners.pop(resource_id, None)
        self._clear_subscription_timer_locked(resource_id)
        self._subscription_ttls.pop(resource_id, None)


def _extract_id_from_event(data: Any) -> str | None:
    """Extract resource ID from an event payload.

    Handles two payload shapes:
      - Wrapper: {sandbox: {id: ...}, ...} → nested resource ID
      - Direct: {id: ...} → top-level ID
    """
    if not isinstance(data, dict):
        return None
    for key in ("sandbox", "volume", "snapshot", "runner"):
        nested: object = data.get(key)  # pyright: ignore[reportUnknownVariableType]
        if isinstance(nested, dict):
            sid: object = nested.get("id")  # pyright: ignore[reportUnknownVariableType]
            if isinstance(sid, str):
                return sid
    top_id: object = data.get("id")  # pyright: ignore[reportUnknownVariableType]
    if isinstance(top_id, str):
        return top_id
    return None

File: libs/sdk-typescript/src/utils/EventSubscriber.ts
/*
 * Copyright 2025 Daytona Platforms Inc.
 * SPDX-License-Identifier: Apache-2.0
 */

import { io, Socket } from 'socket.io-client'

/** Handler receives (eventName, rawData). */
export type EventHandler = (eventName: string, data: any) => void

/**
 * Extracts the resource ID from an event payload.
 *
 * Handles two payload shapes:
 *   - Wrapper: {sandbox: {id: ...}, ...} -> nested resource ID
 *   - Direct: {id: ...} -> top-level ID
 */
function extractIdFromEvent(data: any): string | undefined {
  if (!data || typeof data !== 'object') return undefined
  for (const key of ['sandbox', 'volume', 'snapshot', 'runner']) {
    const nested = data[key]
    if (nested && typeof nested === 'object' && typeof nested.id === 'string') {
      return nested.id
    }
  }
  if (typeof data.id === 'string') {
    return data.id
  }
  return undefined
}

/**
 * Manages a Socket.IO connection to the Daytona notification gateway
 * and dispatches sandbox events to per-sandbox handlers.
 */
export class EventSubscriber {
  private socket: Socket | null = null
  private _connected = false
  private _failed = false
  private _failError: string | null = null
  private listeners = new Map<string, Set<EventHandler>>()
  private registeredEvents = new Set<string>()
  private connectPromise: Promise<void> | null = null
  private ensureConnectPromise: Promise<void> | null = null
  private reconnectAttempts = 0
  private readonly maxReconnectAttempts = 10
  private disconnectTimer: ReturnType<typeof setTimeout> | null = null
  private static readonly DISCONNECT_DELAY_MS = 30_000

  constructor(
    private readonly apiUrl: string,
    private readonly token: string,
    private readonly organizationId?: string,
  ) {}

  /**
   * Idempotent: ensure a connection attempt is in progress or already established.
   *
   * Non-blocking. Fires-and-forgets a connect() call via a stored promise if not
   * already connected and no attempt is currently running.
   */
  ensureConnected(): void {
    if (this._connected) return
    if (this.connectPromise) return
    if (this.ensureConnectPromise) return

    this.ensureConnectPromise = this.connect()
      .catch(() => {
        // Callers check isConnected when they need it
      })
      .finally(() => {
        this.ensureConnectPromise = null
      })
  }

  /**
   * Establishes the Socket.IO connection. Resolves when connected.
   * Throws if the connection fails within the timeout.
   */
  async connect(timeoutMs = 5000): Promise<void> {
    if (this._connected && this.socket) {
      return
    }

    if (this.connectPromise) {
      return this.connectPromise
    }

    this.connectPromise = this.doConnect(timeoutMs)

    try {
      await this.connectPromise
    } finally {
      this.connectPromise = null
    }
  }

  private doConnect(timeoutMs: number): Promise<void> {
    return new Promise<void>((resolve, reject) => {
      // Strip /api suffix to get the origin
      const origin = this.apiUrl.replace(/\/api\/?$/, '')

      const query: Record<string, string> = {}
      if (this.organizationId) {
        query.organizationId = this.organizationId
      }

      this.socket = io(origin, {
        path: '/api/socket.io/',
        autoConnect: false,
        transports: ['websocket'],
        query,
        reconnection: true,
        reconnectionAttempts: this.maxReconnectAttempts,
        reconnectionDelay: 1000,
        reconnectionDelayMax: 30000,
      })

      this.socket.auth = { token: this.token }

      const timer = setTimeout(() => {
        if (!this._connected) {
          this.socket?.disconnect()
          this._failed = true
          this._failError = 'WebSocket connection timed out'
          reject(new Error(this._failError))
        }
      }, timeoutMs)

      this.socket.on('connect', () => {
        clearTimeout(timer)
        this._connected = true
        this._failed = false
        this._failError = null
        this.reconnectAttempts = 0

        // Unref all underlying handles so they don't prevent Node.js process exit.
        // The socket.io connection should not keep the process alive — it's a background
        // enhancement, not a critical resource.
        this.unrefAll()

        resolve()
      })

      this.socket.on('connect_error', (err) => {
        if (!this._connected) {
          clearTimeout(timer)
          this._failed = true
          this._failError = `WebSocket connection failed: ${err.message}`
          reject(new Error(this._failError))
        }
      })

      this.socket.on('disconnect', (reason) => {
        this._connected = false
        if (reason === 'io server disconnect') {
          // Server initiated disconnect - try to reconnect
          this.socket?.connect()
        }
      })

      this.socket.io.on('reconnect', () => {
        this._connected = true
        this._failed = false
        this._failError = null
        this.reconnectAttempts = 0
      })

      this.socket.io.on('reconnect_attempt', () => {
        this.reconnectAttempts++
      })

      this.socket.io.on('reconnect_failed', () => {
        this._connected = false
        this._failed = true
        this._failError = `WebSocket reconnection failed after ${this.maxReconnectAttempts} attempts`
      })

      // Re-register any events that were added before the socket was created
      const pendingEvents = [...this.registeredEvents]
      this.registeredEvents.clear()
      this.registerEvents(pendingEvents)

      this.socket.connect()
    })
  }

  /**
   * Registers Socket.IO event handlers (idempotent -- each event is registered once).
   */
  private registerEvents(events: string[]): void {
    for (const eventName of events) {
      if (this.registeredEvents.has(eventName)) {
        continue
      }
      this.registeredEvents.add(eventName)

      // If socket isn't created yet, the event will be registered when connect() runs
      if (!this.socket) continue

      const handler = (data: any) => {
        const resourceId = extractIdFromEvent(data)
        if (resourceId) {
          this.dispatch(resourceId, eventName, data)
        }
      }

      this.socket.on(eventName, handler)
    }
  }

  /**
   * Registers a handler for events targeting a specific resource.
   * Returns an unsubscribe function.
   *
   * @param resourceId - The ID of the resource (e.g. sandbox ID).
   * @param handler - Callback receiving (eventName, rawData).
   * @param events - List of Socket.IO event names to listen for.
   */
  subscribe(resourceId: string, handler: EventHandler, events: string[]): () => void {
    // Cancel any pending delayed disconnect since we have a new subscriber
    if (this.disconnectTimer) {
      clearTimeout(this.disconnectTimer)
      this.disconnectTimer = null
    }

    // Register any new events with the Socket.IO client
    this.registerEvents(events)

    if (!this.listeners.has(resourceId)) {
      this.listeners.set(resourceId, new Set())
    }
    this.listeners.get(resourceId)!.add(handler)

    return () => {
      const handlers = this.listeners.get(resourceId)
      if (handlers) {
        handlers.delete(handler)
        if (handlers.size === 0) {
          this.listeners.delete(resourceId)
        }
      }

      // Schedule delayed disconnect when no sandboxes are listening anymore
      if (this.listeners.size === 0) {
        this.disconnectTimer = setTimeout(() => {
          if (this.listeners.size === 0) {
            this.disconnect()
          }
        }, EventSubscriber.DISCONNECT_DELAY_MS)
      }
    }
  }

  /** Whether the WebSocket is currently connected */
  get isConnected(): boolean {
    return this._connected
  }

  /** Whether the WebSocket has permanently failed (exhausted reconnection attempts) */
  get isFailed(): boolean {
    return this._failed
  }

  /** The error message if the connection has failed */
  get failError(): string | null {
    return this._failError
  }

  /** Disconnects and cleans up all resources */
  disconnect(): void {
    if (this.disconnectTimer) {
      clearTimeout(this.disconnectTimer)
      this.disconnectTimer = null
    }
    if (this.socket) {
      this.socket.removeAllListeners()
      this.socket.disconnect()
      this.socket = null
    }
    this._connected = false
    this.listeners.clear()
    this.registeredEvents.clear()
  }

  /**
   * Unrefs the underlying TCP socket so the event subscriber doesn't prevent
   * Node.js process exit. This is a no-op in browser environments.
   * Only unrefs the raw socket — does not touch engine.io timers or internals.
   */
  private unrefAll(): void {
    try {
      const engine = (this.socket?.io as any)?.engine
      if (!engine) return

      // Unref the raw TCP socket underneath the WebSocket transport.
      // This tells Node.js not to keep the event loop alive solely for this connection.
      const rawSocket = engine?.transport?.ws?._socket
      if (rawSocket && typeof rawSocket.unref === 'function') {
        rawSocket.unref()
      }
    } catch {
      // Not critical - only affects process exit behavior
    }
  }

  private dispatch(resourceId: string, eventName: string, data: any): void {
    if (!resourceId) return
    const handlers = this.listeners.get(resourceId)
    if (handlers) {
      for (const handler of handlers) {
        try {
          handler(eventName, data)
        } catch {
          // Don't let a handler error break other handlers
        }
      }
    }
  }
}
